### PR TITLE
Changing logging level for config transformation and fixing rule

### PR DESCRIPTION
**Pull request by srikanthjg**
### Description
- Addressed Array Node replacement
- Fixed PipelineDataflowModel transformation from json.
- Addressed relative file path for rules and templates.
- Changed logging level for config transformation
- Fixed rule and template for documentdb
 
### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has a documentation issue. Please link to it in this PR.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-pipeline-parser/src/resources/templates/documentdb-template.yaml [hunk 1]
```diff
@@ -38,16 +38,16 @@
...
           maximum_size: "1mb"
-        aggregate_threshold:
+      aggregate_threshold:
           maximum_size: "128mb"
           flush_capacity_ratio: 0
-        object_key:
-          path_prefix: "${getMetadata(\"s3_partition_key\")}"
-        codec:
+      object_key:
+          path_prefix: "local-test/${getMetadata(\"s3_partition_key\")}"
```
Comment (oeyh) [inline data-prepper-pipeline-parser/src/resources/templates/documentdb-template.yaml:49]: Do we want `local-test` here?

File: data-prepper-pipeline-parser/src/resources/templates/documentdb-template.yaml [hunk 2]
```diff
@@ -38,16 +38,16 @@
...
+      codec:
           json:
 
   "<<pipeline-name>>.s3":
```
Comment (oeyh) [inline data-prepper-pipeline-parser/src/resources/templates/documentdb-template.yaml:53]: Shouldn't have indent here because this is top level pipeline?

**Review by oeyh — COMMENTED** [review]
Is there a way to test this template? The indentations on the two s3 sink configs look off to me. For example, `routes`, `aws`, `bucket`,... all these options currently have the same indent as `s3:`. I don't think that will work.

File: data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/rule/RuleEvaluator.java [hunk 3]
```diff
@@ -54,18 +54,21 @@ public RuleEvaluatorResult isTransformationNeeded(PipelinesDataFlowModel pipelin
...
                 String pipelineJson = OBJECT_MAPPER.writeValueAsString(entry);
                 if (evaluate(pipelineJson, pluginRulesPath)) {
-                    LOG.debug("Rule path {} is evaluated true for pipelineJson {}",pluginRulesPath, pipelineJson);
+                    LOG.info("Rule path {} is evaluated true for pipelineJson {}",pluginRulesPath, pipelineJson);
 
-                    String templateFilePath = transformersFactory.getPluginTemplateFileLocation(PLUGIN_NAME);
-                    PipelineTemplateModel templateModel = yamlMapper.readValue(new File(templateFilePath),
+                    String templateFilePathString = transformersFactory.getPluginTemplateFileLocation(PLUGIN_NAME);
+                    File templateFile = new File(templateFilePathString);
+                    LOG.info("Absolute path of template file: {}",templateFile.getAbsolutePath());
```
Comment (dinujoh) [inline data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/rule/RuleEvaluator.java:68]: We need to fix the path. It refers to local package.

File: data-prepper-pipeline-parser/src/resources/templates/documentdb-template.yaml (hunk 2 above)
Comment (srikanthjg) [reply data-prepper-pipeline-parser/src/resources/templates/documentdb-template.yaml]: Thanks for noticing the indentation, ll fix it.

File: data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/rule/RuleEvaluator.java [hunk 4]
```diff
@@ -54,18 +54,22 @@ public RuleEvaluatorResult isTransformationNeeded(PipelinesDataFlowModel pipelin
...
             try {
                 String pipelineJson = OBJECT_MAPPER.writeValueAsString(entry);
                 if (evaluate(pipelineJson, pluginRulesPath)) {
-                    LOG.debug("Rule path {} is evaluated true for pipelineJson {}",pluginRulesPath, pipelineJson);
+                    LOG.info("Rule path {} is evaluated true for pipelineJson {}",pluginRulesPath, pipelineJson);
 
-                    String templateFilePath = transformersFactory.getPluginTemplateFileLocation(PLUGIN_NAME);
-                    PipelineTemplateModel templateModel = yamlMapper.readValue(new File(templateFilePath),
+                    String templateFilePathString = transformersFactory.getPluginTemplateFileLocation(PLUGIN_NAME);
+                    File templateFile = new File(templateFilePathString);
```
Comment (oeyh) [inline data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/rule/RuleEvaluator.java:68]: Another way to access the file in resources directory is to use `getClass().getClassLoader().getResourceAsStream()` where you can use relative path.

**Review by oeyh — APPROVED** [review]

File: data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/PipelineTransformationConfiguration.java [hunk 5]
```diff
@@ -10,30 +10,34 @@
...
 import javax.inject.Named;
+import java.nio.file.Path;
+import java.nio.file.Paths;
 
 @Configuration
 public class PipelineTransformationConfiguration {
     public static final String TEMPLATES_DIRECTORY_PATH = "TEMPLATES_DIRECTORY_PATH";
-    public static final String RULES_DIRECTORY_PATH = "VALIDATORS_DIRECTORY_PATH";
+    public static final String RULES_DIRECTORY_PATH = "RULES_DIRECTORY_PATH";
+    private static final Path currentDir = Paths.get(System.getProperty("user.dir"));
```
Comment (dinujoh) [inline data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/PipelineTransformationConfiguration.java:20]: `user.dir` is the path where java was run. May not work for all cases.

File: data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/PipelineTransformationConfiguration.java [hunk 6]
```diff
@@ -10,30 +10,34 @@
...
     static String provideRulesDirectoryPath() {
-        return "resources/rules";
+        return currentDir.toString()+ parserRelativePath + "/resources/rules";
     }
 
     @Bean
     @Named(TEMPLATES_DIRECTORY_PATH)
     static String provideTemplateDirectoryPath() {
-        return "resources/templates";
+        return currentDir.toString() + parserRelativePath + "/resources/templates";
```
Comment (dinujoh) [inline data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/PipelineTransformationConfiguration.java:32]: same as above

File: data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/transformer/DynamicConfigTransformer.java [hunk 7]
```diff
@@ -384,17 +397,32 @@ public void replaceNode(JsonNode root, String jsonPath, JsonNode newNode) {
         try {
             if (newNode == null) {
 //                throw new PathNotFoundException(format("jsonPath {} not found", jsonPath));
```
Comment (dinujoh) [inline data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/transformer/DynamicConfigTransformer.java:399]: remove this commented line if not needed.

File: data-prepper-pipeline-parser/src/test/resources/transformation/expected/documentdb2-expected.yaml [hunk 8]
```diff
@@ -26,82 +65,43 @@ documentdb-pipeline-reader:
         username: "${{aws_secrets:secret:password}}"
         password: "${{aws_secrets:secret:password}}"
       s3_prefix: "folder1/folder2"
-  routes:
+  route:
```
Comment (dinujoh) [inline data-prepper-pipeline-parser/src/test/resources/transformation/expected/documentdb2-expected.yaml:68]: both route and routes should work ?

File: data-prepper-pipeline-parser/src/test/resources/transformation/expected/documentdb2-expected.yaml [hunk 9]
```diff
@@ -5,6 +5,45 @@ extension:
...
+  source:
+    s3:
+      codec:
+        json: null
+      scan:
+        buckets:
+          - bucket:
+              name: "bucket-name1"
+              filter:
+                depth: "5"
```
[the remark this answers is not in the document]
Comment (srikanthjg) [reply data-prepper-pipeline-parser/src/test/resources/transformation/expected/documentdb2-expected.yaml:18]: These only test the transformation. At this point, plugins are not loaded. The template that will be used in runtime is in this directory: data-prepper/data-prepper-pipeline-parser/src/resources/templates

File: data-prepper-pipeline-parser/src/test/resources/transformation/expected/documentdb2-expected.yaml (hunk 8 above)
Comment (srikanthjg) [reply data-prepper-pipeline-parser/src/test/resources/transformation/expected/documentdb2-expected.yaml:68]: yes both will work. We could we any field name.
These only test the transformation. At this point, plugins are not loaded. The template that will be used in runtime is in this directory: data-prepper/data-prepper-pipeline-parser/src/resources/templates

File: data-prepper-pipeline-parser/src/resources/templates/documentdb-template.yaml (hunk 1 above)
Comment (srikanthjg) [reply data-prepper-pipeline-parser/src/resources/templates/documentdb-template.yaml]: removed it.

File: data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/PipelineTransformationConfiguration.java (hunk 5 above)
Comment (dlvenable) [reply data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/PipelineTransformationConfiguration.java:20]: Use `data-prepper.dir` to get the runtime directory.

https://github.com/opensearch-project/data-prepper/blob/4f3ab967a14ba23d3957ad6295da0467282cf029/data-prepper-plugins/geoip-processor/src/main/java/org/opensearch/dataprepper/plugins/geoip/extension/MaxMindConfig.java#L51

File: data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/PipelineTransformationConfiguration.java [hunk 10]
```diff
@@ -10,30 +10,34 @@
...
-    public static final String RULES_DIRECTORY_PATH = "VALIDATORS_DIRECTORY_PATH";
+    public static final String RULES_DIRECTORY_PATH = "RULES_DIRECTORY_PATH";
+    private static final Path currentDir = Paths.get(System.getProperty("user.dir"));
+    private static final String parserRelativePath = "/data-prepper-pipeline-parser/src";
 
     @Bean
     @Named(RULES_DIRECTORY_PATH)
     static String provideRulesDirectoryPath() {
-        return "resources/rules";
+        return currentDir.toString()+ parserRelativePath + "/resources/rules";
```
[the remark this answers is not in the document]
Comment (dlvenable) [reply data-prepper-pipeline-parser/src/main/java/org/opensearch/dataprepper/pipeline/parser/PipelineTransformationConfiguration.java:26]: Yes, we should use the classpath.

**Comment by dinujoh** [comment]
Let's clean up the code as follow up PR. I'm fine with merging existing code to test.

**Review by dinujoh — APPROVED** [review]

**Pull request merged**